pkg/storage/ingest/partition_offset_client.go

```diff
@@ -238,3 +249,136 @@ func (p *partitionOffsetClient) FetchTopicLastProducedOffsets(ctx context.Contex
 
 	return offsets, nil
 }
+
+// fetchTopicPartitionsEndOffsets returns the end offset of each partition in the topic. This function returns
+// error if fails to get the offset of any partition (no partial result is returned).
+//
+// The list partitions of the topic is not refreshed for each request, but an in-memory cache is used to reduce
+// the number of requests to Kafka. This is based on the assumption that new partitions don't get added frequently,
+// and a stale view on the existing partitions (but not their offsets) is accepted for a short period of time.
+func (p *partitionOffsetClient) fetchTopicPartitionsEndOffsets(ctx context.Context) (kadm.ListedOffsets, error) {
+	// Fetch the partition IDs. We don't expect partition IDs to change frequently, because we expect
+	// partitions to be pre-provisioned when running Mimir, so use the lookup cache.
+	partitionIDs, err := p.listTopicPartitionIDsWithCache(ctx, 15*time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	list := kadm.ListedOffsets{
+		p.topic: make(map[int32]kadm.ListedOffset, len(partitionIDs)),
+	}
+
+	// Prepare the request to list offsets.
+	topicReq := kmsg.NewListOffsetsRequestTopic()
+	topicReq.Topic = p.topic
+	for _, partitionID := range partitionIDs {
+		partitionReq := kmsg.NewListOffsetsRequestTopicPartition()
+		partitionReq.Partition = partitionID
+		partitionReq.Timestamp = kafkaOffsetEnd
+
+		topicReq.Partitions = append(topicReq.Partitions, partitionReq)
+	}
+
+	req := kmsg.NewPtrListOffsetsRequest()
+	req.IsolationLevel = 0 // 0 means READ_UNCOMMITTED.
+	req.Topics = []kmsg.ListOffsetsRequestTopic{topicReq}
+
+	// Execute the request.
+	shards := p.client.RequestSharded(ctx, req)
+
+	for _, shard := range shards {
+		if shard.Err != nil {
+			return nil, shard.Err
+		}
+
+		res := shard.Resp.(*kmsg.ListOffsetsResponse)
+		if len(res.Topics) != 1 {
```

Comment:
Do we also need to check `for != 0` if the response contains 0 topics it might panic below.

Reply:
I can't understand this comment. If the number of `res.Topics` is different than 1, we return error. So, once we reach the lines below there's the guarantee that `res.Topics[0]` exists. I can't see how it could panic 🤔

Reply:
Ignore me - I read this entirely wrong 🤦